wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,163 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+class TapestryController {
+    const POST_TYPES = array(
+        'TAPESTRY' => 'tapestry',
+        'TAPESTRY_NODE' => 'tapestry_node'
+    );
+
+    /**
+     * Update Tapestry nodes first then
+     * Update the existing Tapestry post if the postId is provided
+     * Otherwise, a new post will be created
+     * 
+     * @param type @post The post data
+     * @param type @postId The postId of the Tapestry
+     */
+    public function updateTapestry($tapestry, $postId = null) {
+        if (!isset($postId))
+            $postId = $this->insertPost($tapestry, 'tapestry');
+        $this->updateNodes($tapestry->nodes, $postId);
+
+        // TODO: Groups and Permisisons data could be added here later
+
+        if (!isset($tapestry->rootId))
+            $tapestry->rootId = $tapestry->nodes[0]->id;
+
+        $tapestry->nodes = $this->getNodeIds($tapestry->nodes);
+        update_post_meta($postId, 'tapestry', $tapestry);
+        return $tapestry;
+    }
+
+    /**
+     * Update Tapestry post's child nodes
+     * 
+     * @param type @nodes An array of child nodes to be updated
+     * @param type @postId The post ID to which the nodes belong
+     * @throws Exception if postId is invalid
+     */
+    public function updateTapestryNodes($nodes, $postId = null) {
+        if (is_null($postId))
+            throw new Exception('postId is invalid');
+        $this->updateNodes($nodes, $postId);
+        return $nodes;
+    }
+
+    /**
+     * Retrieve Tapestry post
+     * 
+     * @param type @postId The postId to be retrieved
+     * @return type Tapestry post data
+     */
+    public function getTapestry($postId) {
+        if (is_null($postId))
+            throw new Exception('postId is invalid');
+        $tapestry = $this->getTapestryById($postId);
+        return $tapestry;
+    }
+
+    private function updateNodes($nodes, $postId) {
+        foreach ($nodes as $node) {
+            if (!isset($node->id)) {
+                $nodePostId = $this->insertPost($node, 'tapestry_node');
+                $metadata = $this->makeMetadata($node, $nodePostId);
+                $node->id = add_post_meta($postId, 'tapestry_node', $metadata);
+            } else {
+                $metadata = get_metadata_by_mid('post', $node->id)->meta_value;
+                $nodePostId = $metadata->post_id;
+            }
+            update_post_meta($nodePostId, 'tapestry_node_data', $node);
+        }
+    }
+
+    private function makeMetadata($node, $nodePostId) {
+        return (object) array(
+            'post_id' => $nodePostId,
+            'title' => $node->title,
+            'coordinates' => (object) array(
+                'x' => $node->fx,
+                'y' => $node->fy
+            )
+        );
+    }
+    
+    private function getTapestryById($postId) {
```

Comment:
Similarly, this is for loading - not for saving, so it doesn't belong in this PR.

Reply:
I removed it.